bpo-29548: Recommend PyObject_Call APIs over PyEval_Call APIs.

methane opened the pull request:
PyEval_Call* APIs are not documented and they doesn't respect PY_SSIZE_T_CLEAN.
So add comment block to recommend them where PyEval_Call APIs are declared.

This commit also changes PyEval_CallMethod and PyEval_CallFunction implementation
same to PyObject_CallMethod and PyObject_CallFunction.

`PyEval_CallFunction(callable, "i", (int)i)` now calls `callable(i)` instead of raising TypeError.

I expect this allows compiler to share some code between PyEval_CallFunction and PyObject_CallFunction.  But I'm not sure.

---
[[bpo-29548](https://bugs.python.org/issue29548)]

JDLH commented:
Per discussion in http://bugs.python.org/issue29510 , I am persuaded that developers should use a global gitignore hide the settings files of their IDE from git and the cpython codebase.  This is a better solution than cluttering up the cpython .gitignore with the settings files of every IDE everywhere. 

For more on global gitignores, see https://help.github.com/articles/ignoring-files/ . 
The gist:
```
% cat > ~/.gitignore_global 
# Eclipse project files
.project
.pydevproject
.settings

# IDE files (ex-Eclipse)
nbproject
.~lock.*
.buildpath
.idea

# OS generated files
.DS_Store
.DS_Store?
._*
.Spotlight-V100
.Trashes
ehthumbs.db
Thumbs.db

# Suggestions for other excludes:
# https://gist.github.com/octocat/9257657
# https://gist.github.com/zpetr/e783d537f8a40bb6ec16
^D

% git config --global core.excludesfile ~/.gitignore_global
```

methane committed:
bpo-29548: Recommend PyObject_Call APIs over PyEval_Call APIs.

PyEval_Call* APIs are not documented and they doesn't respect PY_SSIZE_T_CLEAN.
So add comment block to recommend PyObject_Call* APIs where they are declared.

This commit also changes PyEval_CallMethod and PyEval_CallFunction
implementation same to PyObject_CallMethod and PyObject_CallFunction
to reduce future maintenance cost.  Optimization to avoid temporary
tuple are copied too.

PyEval_CallFunction(callable, "i", (int)i) now calls callable(i) instead of
raising TypeError.  But accepting this edge case is backward compatible.

Carreau commented on Include/ceval.h:
```diff
@@ -7,20 +7,31 @@ extern "C" {
+    PyObject *kwargs) Py_DEPRECATED(3.7);
 
-/* Inline this */
+/* Deprecated from Python 3.7: Use PyObject_CallObject() instead. */
```
did you ment `Use PyObject_Call()` it seem like this patch remove `PyObject_CallObject` ... Autocompleter got a bit overzealous ?

methane replied:
This patch deprecates `PyEval_CallObject`, not `PyObject_CallObject`.

Carreau replied:
Oh, sorry, I got confused with the unified diff and above comment, the proximity of `PyObject_Call()` and `PyObject_CallObject()` made me think you had an `Object` too much in the comment.

methane replied:
There are so many similar call APIs.
That's why I think we should deprecate undocumented and less used APIs.

vstinner commented:
Sorry but I'm confused by the PR title and the commit messages. They don't describe exactly what I see in the code. Can you please update them?

methane commented:
I updated pull request title and body.

ncoghlan reviewed (dismissed):
For the `PyEval_*`/`PyObject_*` functions that are now duplicates, it would be useful to have block comments explaining the equivalence, and the fact the redundant definitions are being kept around for backwards compatibility reasons.

methane commented:
@haypo Would you review this again?

vstinner commented on Objects/call.c:
```diff
@@ -766,30 +766,36 @@ PyEval_CallObjectWithKeywords(PyObject *callable,
 PyObject *
 PyObject_CallObject(PyObject *callable, PyObject *args)
 {
-    return PyEval_CallObjectWithKeywords(callable, args, NULL);
```
I don't understand why you change the code in this commit, and 2 commits later reverts this change. You should practice git rebase -i HEAD~4 to merge the two commits maybe?

vstinner commented on Objects/call.c:
```diff
@@ -766,30 +766,36 @@ PyEval_CallObjectWithKeywords(PyObject *callable,
-    }
-
-    if (!PyTuple_Check(args)) {
+    if (args != NULL && !PyTuple_Check(args)) {
```
Please update the commit message which doesn't make sense compared to the modified code.

This commit fixes a bug, no? It makes sure that a TypeError is raised if kwargs is not a dictionary, right? Please explain it in the commit message.

methane replied:
I remove this bugfix for now and update #87 instead.

methane replied:
done.  I created split issue for #87.

vstinner commented on Objects/call.c:
```diff
@@ -766,30 +766,36 @@ PyEval_CallObjectWithKeywords(PyObject *callable,
 
+    if (args == NULL) {
+        return _PyObject_FastCallDict(callable, NULL, 0, kwargs);
+    }
```
maybe use an else here?

vstinner commented on Objects/call.c:
```diff
@@ -942,22 +948,14 @@ PyObject_CallFunction(PyObject *callable, const char *format, ...)
 PyObject *
 PyEval_CallFunction(PyObject *callable, const char *format, ...)
```
commit message: this change also optimizes PyEval_CallFunction() and PyEval_CallMethod() to avoid temporary tuple when possible.

methane replied:
Main purpose is reducing maintenance cost rather than optimization.
There are very few callers in the world.

I hope compiler can remove duplicated code too, but I'm not sure.

vstinner commented on Include/ceval.h:
```diff
@@ -7,6 +7,10 @@ extern "C" {
 /* Interface to random parts in ceval.c */
 
+/* PyObject_Call(), PyObject_CallFunction(), and PyObject_CallMethod()
+ * are documented and recommended API to call callable objects.
```
I don't understand the "are documented" part, maybe remove it?

I suggest to be more explicit:

" PyEval_CallObjectWithKeywords() is kept for backward compatibility: PyObject_Call(), PyObject_CallFunction() and PyObject_CallMethod() are recommended to call a callable object."

methane replied:
done

vstinner commented:
I'm not confortable yet with GitHub to review commit messages :-/ It's not easy to comment the commit message: I also suggest to add "bpo-29548" to commit messages (at least to one commit message).

methane commented:
> I'm not confortable yet with GitHub to review commit messages :-/ It's not easy to comment the commit message:

I agree. I like Gerrit for this.

> I also suggest to add "bpo-29548" to commit messages (at least to one commit message).

Final commit message is written when pushing "Squash and merge" button.
So I updated pull request title and body.  The title will be first line of commit message, and the body follows.

I hope it works as poor man's Gerrit for a while, until we have new tool to build Misc/NEWS.

methane commented:
I squashed all commit and modified commit message.
